### PR TITLE
Use the same tree for Handshake and Application secrets

### PR DESCRIPTION
**Pull request by raphaelrobert**

### CODE REVIEW

File: src/tree/secret_tree.rs [hunk 1]
```diff
@@ -182,38 +213,63 @@ impl ASTree {
...
+        };
         self.nodes[index_in_tree.as_usize()] = None;
-        self.sender_ratchets[index.as_usize()] = Some(sender_ratchet);
-        application_secret
+        self.handshake_sender_ratchets[index.as_usize()] = Some(handshake_sender_ratchet);
+        self.application_sender_ratchets[index.as_usize()] = Some(application_sender_ratchet);
+        ratchet_secrets
     }
 
     fn hash_down(&mut self, ciphersuite: &Ciphersuite, index_in_tree: NodeIndex) {
```
Comment (kkohbrok) [inline src/tree/secret_tree.rs:248]: More like `derive_down`?

File: src/tree/secret_tree.rs [hunk 2]
```diff
@@ -182,38 +213,63 @@ impl ASTree {
...
         let left_index = left(index_in_tree);
         let right_index = right(index_in_tree, self.size);
-        let left_secret = derive_app_secret(
+        let left_secret = derive_tree_secret(
             &ciphersuite,
             &node_secret,
             "tree",
             left_index.as_u32(),
             0,
             hash_len,
```
Comment (kkohbrok) [inline src/tree/secret_tree.rs:259]: This should be the `KDF.Nh`, but that's a bug in the spec as well. (Same a few lines below.)

File: src/tree/sender_ratchet.rs [hunk 3]
```diff
@@ -79,16 +79,16 @@ impl SenderRatchet {
...
             }
             let secret = self.past_secrets.last().unwrap();
-            let application_secrets = self.derive_key_nonce(&secret, generation, ciphersuite);
+            let ratchet_secrets = self.derive_key_nonce(&secret, generation, ciphersuite);
             self.generation = generation;
-            Ok(application_secrets)
+            Ok(ratchet_secrets)
         }
     }
     fn ratchet_secret(&self, secret: &[u8], ciphersuite: &Ciphersuite) -> Vec<u8> {
```
Comment (kkohbrok) [inline src/tree/sender_ratchet.rs:87]: If the result is a `secret`, it should probably not be `Vec<u8>`, but some struct encapsulating the scret, right? Not sure where this is mean to be used, as it isn't `pub` and it I can't see it being used anywhere here.

**Review by kkohbrok — COMMENTED** [review]
Left a few comments. Also: Do we explicitly delete keys anywhere? I feel like we should and I couldn't find anything here.

File: src/tree/secret_tree.rs (hunk 2 above)
Comment (raphaelrobert) [reply src/tree/secret_tree.rs:259]: Will be fixed with #57 and #46.

File: src/tree/sender_ratchet.rs (hunk 3 above)
Comment (raphaelrobert) [reply src/tree/sender_ratchet.rs:87]: Will be addressed in #94

File: src/framing/mod.rs [hunk 4]
```diff
@@ -344,8 +295,16 @@ impl MLSCiphertext {
             )
             .unwrap();
         let sender_data = MLSSenderData::from_bytes(&sender_data_bytes).unwrap();
-        let application_secrets = astree
-            .get_secret(ciphersuite, sender_data.sender, sender_data.generation)
+        // TODO Handle error
```
Comment (franziskuskiefer) [inline src/framing/mod.rs:298]: TODO

File: src/framing/mod.rs [hunk 5]
```diff
@@ -344,8 +295,16 @@ impl MLSCiphertext {
...
+        let secret_type = SecretType::try_from(&self.content_type).unwrap();
+        // TODO Handle error
```
Comment (franziskuskiefer) [inline src/framing/mod.rs:300]: TODO

File: src/tree/test_secret_tree.rs [hunk 6]
```diff
@@ -0,0 +1,93 @@
+#[test]
+fn test_boundaries() {
```
Comment (franziskuskiefer) [inline src/tree/test_secret_tree.rs:2]: These tests need descriptions what they're supposed to test.

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
I don't know what this PR is doing (no description) but a little drive by review

**Comment by kkohbrok** [comment]
Would it make sense to also add some documentation stating what it does on a high level and how it relates to the standard? I think that would make it easier to grasp the code structure and find bugs, as there is a verbatim description of what is supposed to happen (on a high level) in the comment, which can then be compared to what is happening (on a low level) in the code.

**Comment by raphaelrobert** [comment]
> I don't know what this PR is doing (no description) but a little drive by review

It addresses #72, which in turn is a spec change in the MLS protocol.

**Comment by raphaelrobert** [comment]
> Open issues
> 
>     * There's no description in this PR what it does

We should align how much we need here. We have a flurry of issues that address changes in the MLS protocol spec and they are tagged as such. The changes in the MLS protocol spec are in turn compartmentalized in PRs, making them easy to digest. We can repeat some of the information in the PR itself, if you think that helps.

>     * No description in #72 either

See above.

>     * No link to #72

GH does that automatically when the PR's name says "Closes #XX".

>     * #72 is unassigned

Fixed.

>     * This PR is too big

The limit was reached during review unfortunately. I'll keep a better eye on the size going forward.

**Review by franziskuskiefer — APPROVED** [review]
ok let's get this in and start doing this properly after this one 🙂

**Pull request merged**